USB power state callback, API rework.

eobalski opened the pull request:
This PR is created to resolve:
* point 1. from #13403,
* #20236. 

This PR:
- presents solution for introducing state callback function common to all classes.
- reworks API and hides unnecessary things (refer to #20236). This has to be introduced in order to accomplish #13403.
- adapts samples for new API.

Commit:
usb: Kconfig: Add option for enabling USB device.

Auto enable option allows users to control when to enable USB device.
By default this option is disabled and USB is enabled on system boot up.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
samples: usb: Adapt USB samples for new Kconfig option

The purpose of this commit is to adapt existing examples
for new Kconfig option USB_DEVICE_AUTO_ENABLE.
If configured 'n' then user app needs to enable usb by itself.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

carlescufi commented:
I think that we should have an enum in `usb_device.h`:

```
enum usb_states {
	/** USB error reported by the controller */
	USB_STATE_ERROR,
	/** USB reset */
	USB_STATE_POWERED,
	/** USB connection established, hardware enumeration is completed */
	USB_STATE_DEFAULT,
	USB_STATE_ADDRESS,
	/** USB configuration done */
	USB_STATE_CONFIGURED,
	/** USB connection suspended by the HOST */
	USB_STATE_SUSPENDED,
};
```

Commit:
usb: api: Rework USB device status codes, add USB device state codes.

This commit is combined of two main changes:
- Introducing USB device state enum type based on USB 2.0 spec
  (Chapter 9.)
- Reworking USB device controller status codes to be alligned with
  USB device state and what hardwares need to work properly

By this commit files which use status codes are properly
updated.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
usb: api: Add USB stack state callback.

By this commit global state callback is introduced and its called
whenever stack needs to inform about USB dev. state change.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
usb: api: Clean up for inclusion path. API rework.

By this commit public and private API is reworked.
From user perspective unnecessary things are hidden
by moving them from usb_device.h -> usb_dc.h.

Endpoint related callbacks, structures, etc shall not
be visuble to end user and thus are placed in device
controller header file (usb_dc.h).

Inclusion path is changed so that usb_device.h (public API)
is no longer including (usb_dc.h).

Right now, if user want to implement his own class
or interact direclty with endpoints he needs to include
usb_dc.h.

This commit is solution for #20236.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
usb: api: Add USB device controller state callback.

By this commit usb_device_state_callback is registered
to each USB device controller.

Whenever USB device state is changed callback with proper value
would be issued.

It is up to vendors to implement when callback with each value
should be called.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
usb: samples: Adapt samples for new callback.

By this commit samples are adjust to new API.
When calling usb_enable() usb_state_callback
must be provided. By default NULL pointer is given
to avoid compilation errors.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

pabigot commented on drivers/console/uart_console.c:
```diff
@@ -602,7 +602,7 @@ static int uart_console_init(struct device *arg)
 	int ret;
 
-	ret = usb_enable();
+	ret = usb_enable(NULL);
```
I believe all these need to be in the commit where the signature to `usb_enable()` changes so that the tree builds at every commit.

emob-nordic commented:
AFter discussion with @pabigot this could be acheived with smaller PR. Closing this one.
